### ISSUE TITLE
Add support for Data API (AWS Aurora Serverless)

### ISSUE THREAD

**Issue by schickling** [prisma/prisma2#1964]
## Problem

Prisma currently doesn't seem to support the [Data API for AWS Aurora Serverless](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html).

## Solution

It would be great to add support for it which is especially useful for applications running on serverless environments like AWS Lambda.

**Comment by d2kx** [comment]
I absolutely love this and wanted it for some time, however be mindful that there are some limitations right now, especially for those of us in the EU:

- Aurora Serverless for MySQL only supports MySQL 5.6 right now, which I am not sure is still supported by Prisma 2 (the supported-databases.md only mentions MySQL 5.7). Aurora Serverless for MySQL 5.7 will likely launch this year, but it hasn't so far

- Aurora Serverless for PostgreSQL only supports PostgreSQL 10.7 right now, and is also only available in Ireland, not Frankfurt (Germany) in the EU so far

- Data API for Aurora Serverless is only available in some regions so far, e.g. not in Frankfurt (Germany) (EU-Central-1), but in Ireland (EU-West-1)

**Comment by medv** [comment]
Linking my comment for relevance, as suggested: https://github.com/prisma/prisma/issues/2133#issuecomment-635062736

Will love to do any testing when needed.

**Comment by d2kx** [comment]
A few updates since my comment above.

- Data API for Aurora Serverless is now available in more regions, including eu-central-1 (EU/Frankfurt), see https://aws.amazon.com/about-aws/whats-new/2020/05/amazon-rds-data-api-and-query-editor-available-additional-regions/

- Aurora Serverless for PostgreSQL (10.7) is now available in more regions, including eu-central-1 (EU/Frankfurt). It includes support for Data API for Aurora Serverless

**Comment by MaxAst** [comment]
I would love for this to exist. Especially in a serverless setup, this would be a great way to avoid the connection limit issue. Is there a more detailed RFC or something detailing how the prisma team would like this to be implemented?

**Comment by Albert-Gao** [comment]
May I have a bold idea here:

In terms of supporting the Data API, can we compile the whole GraphQL query into one SQL and execute it over Data API?

What a performance...

**Comment by hemildesai** [comment]
I'm using https://redwoodjs.com/ which deploys a graphql API as a serverless function. It would be great to have Data API support so that AWS Aurora Serverless can be used with such serverless applications.

**Comment by matthewmueller** [comment]
More use cases described in this issue: https://github.com/prisma/prisma/issues/2133

**Comment by justinhandley** [comment]
I'm incredibly interested in the possibility of making this happen.  Has anyone looked deeply at what it would take yet?  I use Prisma all the time but have yet to actually go through the source code.  If this is a manageable project it may be one I want to take on, but I'm wondering if anyone has looked into scope, or if Prisma has plans already in place for adding this?

**Comment by f0rr0** [comment]
There is Data API client which is also used by TypeORM https://github.com/ArsenyYankovsky/typeorm-aurora-data-api-driver

The Data API seems to be plagued with some issues related to casting types like json, dates etc which is handled by the above client https://github.com/jeremydaly/data-api-client/pull/57

It seems like to be able to use this Prisma would have to explicitly provide the types to the Data API client.

@medv mentioned that they are using this with TypeORM and can probably share their experience.

**Comment by emm-jay** [comment]
@medv , would love a bit of expansion on why you ditched typeorm/typeorm-aurora-data-api-driver in favor of data-api-client.

**Comment by medv** [comment]
TypeORM itself did not fit our team, so our decision was less about the data api implementation. `data-api-client` allowed us to retain our temporary ORM solution in the back-end as well as provide very low latency and small bundle size for our serverless front-end served from lambda@ledge (temporary because we are eagerly awaiting prisma with migrate and data-api to be well defined and widely used).

**Comment by thdxr** [comment]
Is a data api driver something that is actively being worked on?

**Comment by justinhandley** [comment]
So, I asked a prisma developer this directly, and he took the question to the product team.  I specified that I had a project launching in July, and if the integration would be done by summer, we would build with prisma.  They said that currently they have no plans to prioritize this, and it would not likely be done this year.  Obviously the more interest the more likely they are to prioritize, but right now I don’t think it is really on the radar.

**Comment by cmonteiro128** [comment]
Also would be interested in this

**Comment by thdxr** [comment]
> So, I asked a prisma developer this directly, and he took the question to the product team. I specified that I had a project launching in July, and if the integration would be done by summer, we would build with prisma. They said that currently they have no plans to prioritize this, and it would not likely be done this year. Obviously the more interest the more likely they are to prioritize, but right now I don’t think it is really on the radar.

That's extremely disappointing.  @justinhandley what are you planning on doing?

**Comment by softmarshmallow** [comment]
can i use aurora serverless as prisma db or not? (or is this thread only about using data api for internal processing?)
https://www.prisma.io/docs/reference/database-reference/supported-databases

'cause aurora serverless is listed as supported db

**Comment by justinhandley** [comment]
@softmarshmallow, you can use aurora serverless.  This conversation is about the Data API.  In the fantasy of moving to a completely scalable and serverless environment, as well as having your database scalable, your endpoints would ideally also be scalable and served from edge servers.  Unfortunately, with aurora serverless, you can’t access the database from outside of the region the database is in, so you can’t have an api served on edge lambdas connect.  You could theoretically achieve this with the aurora data api.  But, you are good to go.

**Comment by jan-wilhelm** [comment]
@schickling himself opened this issue so I'm sure he sees the huge opportunity this would mean for developers using Prisma. In this whole JS world that is trending more and more towards Serverless, scalable database access is one of the largest bottlenecks. I believe Prisma would benefit hugely from this move and I wonder about the reasoning is to not offer support in the short and medium term.

Disclaimer: I do have to say though that I do not know how resource-intensive it would be to make these changes (time-wise). From an outsider perspective, wouldn't it "just" mean sending the SQL strings (that the engine already generates) via HTTP rather than a native connection? Looking forward to a quick response, if possible at all :)

**Comment by f0rr0** [comment]
On a separate topic but kind of related, I think it will be in the best interest of developers using Prisma to mention the current status of Data API and RDS Proxy compatibility in the docs and wherever else applicable. Right now this information is painfully hard to weed out. I have opened a PR for the same https://github.com/prisma/docs/pull/1409

**Comment by sorenbs** [comment]
@medv - I'm wondering if you could share your experience using the Aurora Serverless Data API so far? I have talked to a number of people who had trouble using it at scale reporting issues such as large increase in response time, random failures and sometimes receiving wrong data. Have you experienced something like that or has it been smooth sailing?

**Comment by medv** [comment]
> @medv - I'm wondering if you could share your experience using the Aurora Serverless Data API so far? I have talked to a number of people who had trouble using it at scale reporting issues such as large increase in response time, random failures and sometimes receiving wrong data. Have you experienced something like that or has it been smooth sailing?

I can definitely confirm all that you said. I wonder how Aurora Serverless 2 would fare and looking at fauna in the meantime, but we have opted for a non-serverless Aurora with 1000 connection limit as a stopgap. Slight greeting teeth at the tape holding this all together. The idea is incredible from the getgo and in theory this is what we wish to do, but we have to bide our time with these interim limbo dances in the face of reality. Our usecase is not massive, but it's still about ~5mill users a month so plenty for us to invalidate a lot of approaches which we have so far.

1. Prisma on top of Data API would solve massive dev ux overhead, but may not entirely solve Aurora Serverless issues (may address some client layer issues over time due to sheer number of users and amazing prisma dev team having some ownership). This would already be the front-running solution in our eyes.
2. Fauna with a prisma binding would be pretty ace, but probably has unforseen issues in of itself, I will be happy to report back after our experiments which are on the cards but have no real date at of now.
3. Ultimately, Prisma + whatever database driver that addresses connection pools would win the internet in our eyes after having played with a lot of arrangements.

**Comment by sorenbs** [comment]
Thank you @medv - that is extremely helpful!

One of our product teams is tasked with solving the problems you are describing. I would love to get on a call together with our pm to make sure we capture  as much as possible. Would you be open to a call sometime in the coming weeks? You can contact me directly at schmidt@prisma.io

**Comment by mikestopcontinues** [comment]
How about opening up the connector interface so that others could contribute? Though I wouldn't turn down direct support either. :)

**Comment by sorenbs** [comment]
@revmischa - you might be interested in talking to the team working on the new Data Proxy feature that is aiming to solve this problem. If you talk to the PM, we can make sure to design it in a way that meets your needs: https://prisma103696.typeform.com/to/IG9ZKSKk

**Comment by albertoperdomo** [comment]
Yes, I'd love to talk to you about your use case @revmischa. You can schedule time with me using this: https://calendly.com/alberto-prisma/45-min-zoom-call

**Comment by daverickdunn** [comment]
@revmischa I'm in the same boat. The CDK is absolutely amazing, all it needs is a good ORM to integrate with Aurora Serverless and we're bordering on cloud perfection. TypeORM and the driver you linked work, but they're both full of nasty surprises and workarounds. I asked for an update on this in the Prisma Slack less than 2 weeks ago and was told (and I believe quite authoritatively) that the Data API is orthogonal to how Prisma is designed and so it ain't on the roadmap, at least not for the foreseeable future. 

That said, I think in a matter of months the CDK's popularity will have reached a level that the Prisma team will want to get onboard. Hopefully we have to stick out TypeORM for only a little longer...

**Comment by revmischa** [comment]
> @revmischa I'm in the same boat. The CDK is absolutely amazing, all it needs is a good ORM to integrate with Aurora Serverless and we're bordering on cloud perfection. 
I've been building out some reusable constructs for Prisma in CDK: https://github.com/jetbridge/jetkit-cdk/pull/13

**Comment by victormier** [comment]
So basically Prisma wants us to use their (paid) solution, which requires spinning up a proxy in an EC2 instance (if I'm correct), instead of leveraging what Aurora Serverless supports natively. I guess this is a business decision and not a technical one.

I might be wrong.

**Comment by revmischa** [comment]
The 500ms setup time for a postgres SSL connection with prisma is really hurting my lambda cold start performance. Combined with loading the prisma depencies it adds about 1.5 seconds to all of my lambda functions that use prisma. It's a terrible user experience.
If a "thin" version of the prisma client could be built and used the aurora data API it would solve all my problems. Also it would save me money from having to put my lambdas in a VPC with my database, and paying for NAT gateways.

**Comment by daviddripps** [comment]
Just checking in on this. Would love to use Prisma for my next project, but I'm all in on the AWS stack.

**Comment by collinscangarella** [comment]
> So basically Prisma wants us to use their (soon to be paid) solution, which requires spinning up a proxy in an EC2 instance (if I'm correct), instead of leveraging what Aurora Serverless supports natively. I guess this is a business decision and not a technical one.
> 
> I might be wrong.

The data API is a solution for those using aurora serverless while the prisma data proxy solution is a solution regardless of vendor. Not to mention ease of implementation on prisma's end. Sounds like an easy decision to me.

**Comment by revmischa** [comment]
Using Prisma on AWS would be vastly easier and cheaper if I could use the Data API. The API is basically sending a SQL query in an API call. I hope this is possible some day with Prisma.

**Comment by justinhandley** [comment]
I had a conversation with the Prisma crew over a year ago about this, and just to frame the conversation - they said it is a low priority due to the amount of work.  Prisma is a Rust-based service that integrates with databases.  The amazon Data API is not a database, it is a 3rd party API - which means that everything about how Prisma works would have to be redone to integrate with it. On the flip side, Amazon could choose to make Aurora Serverless  (which Prisma can connect to because it is actually a database) available to public connections so that you could run Prisma on edge lambdas and achieve total 'serverless' infrastructure, but they aren't doing that either.

**Comment by DavidChouinard** [comment]
Aurora Serverless v2 allows public connections

**Comment by thdxr** [comment]
If you definitely need this I'd recommend using https://github.com/koskimas/kysely/

We wrote a data api adapter here https://github.com/serverless-stack/kysely-data-api

Prisma is a no go for any serious serverless work so even without data api it's not a good fit

**Comment by revmischa** [comment]
> Out of curiosity: How standard is that SQL? Is it more like MySQL or PostgreSQL? Or ANSI? How standard is the returned data? That https://github.com/jeremydaly/data-api-client exists indicates it is quite different, and needs elaborate transformation to be more usable?

The SQL is the same as the query you would run with libpq or with mysql's client lib. It's running a query on your DB just without the low-level networking part of it. If prisma is capable of building the parameterized SQL query, it should be relatively straightforward to send the query via API rather than the native pg or mysql library. The response shape is a list of typed values - see Dax's example above https://github.com/serverless-stack/kysely-data-api/blob/master/src/data-api-driver.ts#L108

The kysely data API driver is a very nice and minimal implementation, thanks @thdxr

**Comment by janpio** [comment]
Thanks @revmischa for that context. Seems we should hack on a prototype and see where we end up. The possible benefits are pretty clear.

**Comment by mariomnts** [comment]
Do you know if this feature is included at least in the Prisma's roadmap?

**Comment by thdxr** [comment]
> > Do you know if this feature is included at least in the Prisma's roadmap?
> 
> Aurora Serverless v1 is reaching its [End Of Life ](https://repost.aws/questions/QUdyFm-B78QRCFNNuJTihntw/serverless-amazon-aurora-my-sql-1-end-of-life-on-february-28-2023) on Februray 28, 2023. And as long as I know Aurora Serverless v2 [does not support data api client ](https://stackoverflow.com/questions/72231268/when-will-aurora-serverless-v2-have-a-data-api).

I don't think serverless v1 is reaching EoL

**Comment by janpio** [comment]
The thread link clarified that only "Amazon Aurora MySQL-Compatible Edition version 1" (which means MySQL 5.6) on Aurora Serverless v1 is EOL. "Aurora MySQL version 2" and "Aurora MySQL version 3" are still supported on "Aurora Serverless v1" and will be.

**Comment by lefnire** [comment]
@nickygb I didn't realize v2 doesn't support Data API (which is definitely confirmed after deep-diving those threads, I guess I was inadvertently using v1 via SST). That being so, wouldn't Prisma be *easier* to implement for v2? It seems v2 uses traditional connection-handling, with connection-pooling via RDS Proxy and such. Has anyone tried just using Prisma as-is, connecting to an RDS Proxy -> Serverless V2, and see if it just works?

**Comment by janpio** [comment]
Prisma fully supports AWS Aurora Serverless (any version) as long as you use the traditional connection based connection string and not the Data API. (RDS Proxy is _another_ service with its own challenges: https://github.com/prisma/prisma/issues/5866)

**Comment by lefnire** [comment]
Thanks for the reply. I really should have come to understand the difference between Data API & Aurora Serverless before going down this journey.

**Comment by thdxr** [comment]
@lefnire 

The reason we're sticking to v1 in SST is because it is the only viable way to use RDS and be VPC-less.  That's a constraint we find important for building fully serverless systems for many reasons.

v2 isn't necessarily "better" - it's just choosing a different set of tradeoffs and we haven't found them compelling enough to move off of v1

additionally Prisma's issues isn't directly with RDS - it's with Lambda. Prisma does not run well in Lambda due to its architecture

**Comment by lefnire** [comment]
> Prisma does not run well in Lambda due to its architecture

That's good to know, as I was just about to pros/cons an investigation on migrating my SST app to aurora v2. I'm already using a VPC for EFS in my SST app, so I'm fine with that (though I understand why you want to avoid that need for most users). I'll do some digging around the Lambda issues so I'm not pestering you here.

The main reason I was considering v2 is to use Prisma. I've been banging my head against kysely-data-api for a while. I'll chime in there if/when I can, but I just switched to raw SQL queries everywhere so I can launch fast. Even so I hit Data API snags, like no array support (`IN (..)`, `ARRAY[...]`, etc - https://github.com/aws/aws-sdk/issues/9). This isn't the first such snag, so I'd rather take the VPC/Nat/Proxy hit than gotchas. But if the issue is more around Lambda usage.. anyway, I'll do some digging around. Thanks for the reply!

**Comment by LinusU** [comment]
> Prisma does not run well in Lambda due to its architecture

Is there any more information about this? Is this an official stance?

We are running Prisma on Lambda and haven't ran into any problems so far, but I'm really interested in reading more about this...

**Comment by janpio** [comment]
No, this is not an official stance and we do not agree with it. Prisma has its rough corners, especially in serverless environments, but those most often are not hard blockers and we are working on improving them.

If you want to discuss problems Prisma supposedly has with Lambda, please do so in a new discussion and not this feature request for AWS Data API support. Thanks.

**Comment by thdxr** [comment]
@LinusU 

Prisma requires large binaries to be shipped with your lambda function and started up on coldstarts.  These add significant cold start time in lambda environments (upwards of 2 seconds) - this is well documented in several places.  I find it kind of odd prisma is suggesting these are "supposed issues" and it is incredibly relevant to this issue because it's the entire reason why people want direct data api support

**Comment by revmischa** [comment]
My primary motivation for using Data API is so that I can skip provisioning a VPC (slows down deployments, more complicated infra) and NAT gateways ([money](https://www.youtube.com/@e-40-revenueretrievin4albu6/playlists))

Of course if it can reduce cold starts in lambda that is critically important as well for us

AWS [says](https://repost.aws/questions/QUc6tQ-_TsRWeSXAxSlqTPEg/is-data-api-for-aurora-v-2-v-3-planned) Data API support in Aurora Serverless V2 is an internal feature request with no ETA, I haven't the foggiest notion what that signifies in relation to temporal concerns of the flesh

**Comment by sorenbs** [comment]
Thus issue contains multiple mentions of Prisma ORM not working well in Lambda - specifically by causing slow cold-starts.

This blogpost outlines work done over the past 9 months to improve Prismas startup performance, which benefits all users, but especially when deploying to serverless functions: https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t

Prismas startup performance is now on par with the best other ORMS, and faster than most of our competitors. Establishing a database connection still adds to the overall cold-start time, but Prisma adds no significant overhead.

**Comment by raymondKelly** [comment]
> This issue contains multiple mentions of Prisma ORM not working well in Lambda - specifically by causing slow cold-starts.
> 
> This blogpost outlines work done over the past 9 months to improve Prismas startup performance, which benefits all users, but especially when deploying to serverless functions: https://www.prisma.io/blog/prisma-and-serverless-73hbgKnZ6t
> 
> Prismas startup performance is now on par with the best other ORMS, and faster than most of our competitors. Establishing a database connection still adds to the overall cold-start time, but Prisma adds no significant overhead.

I can confirm this! I recently did a POC of TypeORM, MikrORM and Prisma in an AWS Lambdas via AWS CDK against Aurora Serverless V2 and the performance both with cold and warm starts was the same or better when using Prisma. The only remaining concern I have is with the large lambda size when using Prisma in comparison to the other ORMs which in theory should cause slower cold starts but I did not see that being an issue in my testing.

**Comment by janpio** [comment]
(We finally have to conclude our investigation and do a blog post, until then @raymondKelly:  That matches our observation. As far as we can see, the problem is not "big lambda size" but "big lambda size because of many tiny files or lots of JavaScript that has to be parsed". Foruntately, our query engine only makes the Lamba bigger because of that one native code file, and hence usually does not impact cold start negatively.)

**Comment by janpio** [comment]
Hey, we at Prisma have been working on something related for the past few weeks. We might have something to share and test soon.
If you are interested in helping us, please respond to this super short survey and leave your email address so we can get in touch with instructions: https://pris.ly/survey/driver-adapter-aws-data-api Thanks.

**Comment by raymondKelly** [comment]
@janpio AWS just released RDS Data API support for Aurora serverless V2 https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-aurora-postgresql-rds-data-api/ would it make sense that I open up another ticket for that?

Is there any update on the result of the survey linked and any update on if this adapter has made it onto the roadmap? I would think this is maybe the most popular serverless DB being used.

**Comment by janpio** [comment]
> AWS just released RDS Data API support for Aurora serverless V2 [aws.amazon.com/about-aws/whats-new/2023/12/amazon-aurora-postgresql-rds-data-api](https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-aurora-postgresql-rds-data-api/) would it make sense that I open up another ticket for that?

I think with [AWS retiring Aurora Serverless v1, and the Data API connected to that](https://www.infoq.com/news/2024/01/aurora-serverless-v1-retirement/), this issue automatically becomes a feature request for adding support for the Data API based on the now only available option with Aurora Serverless v2 - so no issue necessary really.

The survey got a small number of responses for AWS Data API, <4%. PlanetScale and Neon are much more popular in our userbase (>20% of responses each). So although I recognize that some really want to use AWS Data API, this does not put it near the top of our priorities unfortunately. 

Driver adapters of course can be written by anyone, so maybe someone from the community wants to pick this up and figure out if this is possible.

**Comment by medv** [comment]
It's worth pondering that if Data API was supported since 2020, those survey results would look really different and those other database services themselves may have actually been less popular overall as a result, because of how many serverless prisma consumers there would have been.

**Comment by raymondKelly** [comment]
@janpio Understandable, thank you for your reply. My only feedback would be that Aurora Serverless on AWS is likely to have a much larger user base in the future than the others mentioned and Prisma I would guess that Prisma would be strategically best positioned in aligning support for that user base so as to grow your own user base.

**Comment by mariomnts** [comment]
@janpio Is there any good tutorial or documentation out there about writing Prisma driver adapters? Thanks!

**Comment by mariomnts** [comment]
> No, they are super fresh and no documentation yet. Looking at our codebase (https://github.com/prisma/prisma/tree/main/packages, `adapter-*`) should give you an easy understanding though. They are basically a defined API to be exported, and then just the internal are different for each driver adapter - calling the underlying database client or driver. If you have more questions, maybe start a discussion in this repo and I will try to be helpful (feel free to ping me when creating it).

This is a helpful starting point, thanks!!

**Comment by raymondKelly** [comment]
@mariomnts Pls let me know if you decide to work on an adapter as I may be interested in helping.

**Comment by mariomnts** [comment]
> @mariomnts Pls let me know if you decide to work on an adapter as I may be interested in helping.

No, I haven't looked for the time, it's a stressful time of the year to me. I can do as @janpio suggested and create a discussion as I hace some questions regarding the internals of the adapters though

**Comment by mariomnts** [comment]
Discussion created: https://github.com/prisma/prisma/discussions/22721

**Comment by zdenham** [comment]
Would still love to see this supported, and would give a modest sponsorship to anyone who works on a AWS Data API driver. I've been told to port to Drizzle but I just can't seem to get behind their api.

**Comment by raymondKelly** [comment]
@janpio I am currently putting together a POC for this. Is there anyone at Prisma I could touchbase with to ask questions? I have a project setup via CDK that stands up an Aurora serverless database and also creates a Lambda that uses the RDS Data API client to perform queries and transactions. I am now starting to build the adapter and following the patterns set forth via Neon and Planetscale but I have some questions.

**Comment by kevinlinar** [comment]
2026, supporting the implementation of this feature.
